core/ops.rs

```diff
@@ -13,14 +14,28 @@ pub type Buf = Box<[u8]>;
 pub type OpAsyncFuture<E> =
   Pin<Box<dyn Future<Output = Result<Buf, E>> + Send>>;
 
-pub(crate) type PendingOpFuture =
+pub(crate) type PendingOpInnerFuture =
   Pin<Box<dyn Future<Output = Result<(OpId, Buf), CoreError>> + Send>>;
 
+pub(crate) struct PendingOp(pub PendingOpInnerFuture, pub bool);
+
+impl Future for PendingOp {
+  type Output = Result<(OpId, Buf), CoreError>;
+  fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+    self.as_mut().0.as_mut().poll(cx)
+  }
+}
```

Comment:
Is this really necessary? Seems odd.

Reply:
Agree, I'm still advocating for `Isolate.pending_optional_ops` field. 

Then in `Isolate::poll`:
```rust
...
loop {
  inner.have_unpolled_ops = false;
  #[allow(clippy::match_wild_err_arm)]
  match inner.pending_ops.poll_next_unpin(cx) {
    Poll::Ready(Some(Err(_))) => panic!("unexpected op error"),
    Poll::Ready(None) => break,
    Poll::Pending => break,
    Poll::Ready(Some(Ok((op_id, buf)))) => {
      let successful_push = inner.shared.push(op_id, &buf);
      if !successful_push {
        // If we couldn't push the response to the shared queue, because
        // there wasn't enough size, we will return the buffer via the
        // legacy route, using the argument of deno_respond.
        overflow_response = Some((op_id, buf));
        break;
      }
    }
  }
  #[allow(clippy::match_wild_err_arm)]
  match inner.pending_ops_optional.poll_next_unpin(cx) {
    Poll::Ready(Some(Err(_))) => panic!("unexpected op error"),
    Poll::Ready(None) => break,
    Poll::Pending => break,
    Poll::Ready(Some(Ok((op_id, buf)))) => {
      let successful_push = inner.shared.push(op_id, &buf);
      if !successful_push {
        overflow_response = Some((op_id, buf));
        break;
      }
    }
  }
}
...
```

Reply:
@bartlomieju But isn't that having the chance of starving?

Reply:
@kevinkassimo you might be right, then I guess using [`select`](https://docs.rs/futures/0.3.1/futures/stream/fn.select.html) would be in order. 

Then final check returning `Poll` will still check if `pending_ops` is not empty ignoring `pending_ops_optional`